polynote-frontend/polynote/messaging/receiver.ts

```diff
@@ -326,21 +324,39 @@ export class NotebookMessageReceiver extends MessageReceiver<NotebookState> {
                         }
                     })
                     .when(messages.InsertCell, (g, l, cell: NotebookCell, after: number) => {
-                        const newCell = this.cellToState(cell);
-                        const insertIdx = s.cellOrder.findIndex(id => id === after) + 1;
+                        const insertIdx = (s.cells[after]?.idx ?? -1) + 1;
+                        const newCell = this.cellToState(cell, insertIdx);
+                        // first, update all cell indices
+                        const cellUpdates = Object.values(s.cells).reduce<Record<number, UpdateLike<CellState>>>((acc, state) => {
+                            if (state.idx >= insertIdx) {
+                                acc[state.id] = setProperty("idx", state.idx + 1)
+                            }
+                            return acc
+                        }, {})
+                        // add the new cell
+                        cellUpdates[cell.id] = setValue(newCell)
                         return {
-                            cells: {
-                                [cell.id]: setValue(newCell)
-                            },
-                            cellOrder: insert(newCell.id, insertIdx)
+                            cells: cellUpdates,
                         }
                     })
                     .when(messages.DeleteCell, (g, l, id: number) => {
-                        const idx = s.cellOrder.indexOf(id)
-                        if (idx > -1) {
+                        const deleteIdx = s.cells[id]?.idx;
+                        // Remove deleted cell and update other indices as necessary.
+                        const updatedCells = Object.values(s.cells).reduce<Record<number, CellState>>((acc, state) => {
+                            // no change to cells above the deleted one
+                            if (state.idx < deleteIdx) {
+                                acc[state.id] = state;
+                            }
+                            if (state.idx >= deleteIdx && state.id !== id) {
+                                acc[state.id] = {...state, idx: state.idx - 1}
+                            }
+                            return acc
+                        }, {})
+                        if (deleteIdx !== undefined) {
                             return {
-                                cells: removeKey(id),
-                                cellOrder: removeIndex(s.cellOrder, idx),
+                                // TODO: This sets all the cell values, but ideally it would be `diff` update of only what changed.
+                                //       I couldn't figure out how to remove a key and update other keys at the same time.
+                                cells: setValue(updatedCells),
```

Comment:
FYI @jeremyrsmith  - I wasn't able to delete a key and update others within the same object using an update, so I ended up having to just brute force with a `setValue` here.